### DISCUSSION TITLE
Filament Sensor

### DISCUSSION THREAD

**Discussion by tjhayko** [bassamanator/Sovol-SV06-firmware#197] [General]
I've had some issues with the filament sensor with my last print on my SV06 plus with BTT Pad 7 running Klipper.

Fortunately, it was the first time I was getting close to not having quite enough filament left on the roll to finish a print.  

I ran out of filament, but my print didn't pause when I did.  I've done a quick check using the console, and the sensor appears to be working, as it correctly detects when filament is in and not in the sensor.

is there any additional configuration that I should be done to make the print pause when I run out of filament?

I'm getting a few nearly empty rolls of filament that I'd like to use up.

Thanks in advance

**Comment by bassamanator** [comment]
Please see [this section](https://github.com/bassamanator/Sovol-SV06-firmware/tree/master?tab=readme-ov-file#what-are-some-settings-that-i-can-change), and enable the filament sensor usage.

**Comment by toeachbyeach** [comment]
Sorry to be annoying but I have the same issue. When I run:

QUERY_FILAMENT_SENSOR SENSOR=filament_sensor

I get a response that: 

Filament Sensor filament_sensor: filament detected

Which is accurate. My problem seems to stem from the command for pause not running.

I have had a good look through settings and configs and come up short. 

Is there a way to manually send the same command/script to pause on runout?

**Reply by bassamanator** [reply]
See my previous post on how to enable the filament sensor. You need not change/write anything else to make the filament sensor pause on runout, as can be seen [here](https://github.com/bassamanator/Sovol-SV06-firmware/blob/9b025cde7309a1a51f9a46407e279d0636f8f121/printer.cfg#L224).

**Reply by bassamanator** [reply]
Let me know how it goes.

**Reply by toeachbyeach** [reply]
Firstly, amazing work. You are the single reason I am not struggling with my printer and the hobby generally. 

I forgot to mention I am on an SV06+ so my printer.cfg is the below as per your github printer.cfg for the SV06+:

[filament_switch_sensor filament_sensor]
switch_pin: PA4
pause_on_runout: True
insert_gcode:
    { action_respond_info("Insert Detected") }
runout_gcode:
    { action_respond_info("Runout Detected") }
    CONDITIONAL_BEEP i=3 dur=300 freq=400

Should it be this instead?

[filament_switch_sensor filament_sensor]
switch_pin: !PA4 # "Pulled-high"
pause_on_runout: True
insert_gcode:
    M117 Insert Detected
    { action_respond_info("Insert Detected") }
runout_gcode:
    M117 Runout Detected
    { action_respond_info("Runout Detected") }
    CONDITIONAL_BEEP i=3 dur=300 freq=400

**Reply by bassamanator** [reply]
What is found on this repo has been *tested*, and found to be correct, either by myself, or by multiple other users. You are welcome to switch things up however if they aren't working for you.

To test out the sensor do the following.
1. remove filament from nozzle.
2. cut a short piece of filament, let's say 6 inches, and insert it into the sensor.
3. start a print and let it `print air`
4. remove the filament from the sensor
5. the print should pause.
6. insert the filament back into the sensor, and hit resume in your dashboard
7. the print should resume

See how that goes.

**Comment by toeachbyeach** [comment]
Hi there, thanks for the reply. Sorry I have been busy.

I have now tried as you said and nothing happens. If I query the sensor in the console with:

QUERY_FILAMENT_SENSOR SENSOR=filament_sensor

I get the correct response with and without the filament in the sensor.

**Reply by bassamanator** [reply]
Let's see your `[gcode_macro _globals]` section.

**Reply by toeachbyeach** [reply]
```
[gcode_macro _globals]
variable_filament_sensor_enabled: 1 # NOTE Enable(1) or disable(0) the filament sensor, if one is connected
variable_beeping_enabled: 1 # NOTE Enable(1) or disable(0) beeping everywhere except during gantry calibration
variable_bed_temp_over: 10 # NOTE Start print if bed temperature is over by this amount, otherwise wait for temperature drop
variable_kamp_enable: 1 # NOTE Enable(1) or disable(0) KAMP (adaptive mesh)
variable_pre_purge_prime_length: 1.40
gcode:
    # Don't delete this section

[gcode_macro CONDITIONAL_BEEP]
gcode:
    # Parameters
    {% set i = params.I|default(1)|int %}           ; Iterations (number of times to beep).
    {% set dur = params.DUR|default(100)|int %}     ; Duration/wait of each beep in ms. Default 100ms.
    {% set freq = params.FREQ|default(2000)|int %}  ; Frequency in Hz. Default 2kHz.

    {% set BEEPING_ENABLED=printer["gcode_macro _globals"].beeping_enabled|default(-1)|int %}

    {% if BEEPING_ENABLED == 1 %}
        BEEP I={i} DUR={dur} FREQ={freq}
    {% endif %}

[gcode_macro ADJUST_FILAMENT_SENSOR_STATUS]
gcode:
    # Parameters
    {% set NEWSTATUS = params.ENABLE|default(-1)|int %}

    {% set FILAMENT_SENSOR_ENABLED=printer["gcode_macro _globals"].filament_sensor_enabled|default(-1)|int %}

    {% if FILAMENT_SENSOR_ENABLED == 1 and NEWSTATUS != -1 %}
        SET_FILAMENT_SENSOR SENSOR=filament_sensor ENABLE={NEWSTATUS}
    {% endif %}
```

**Reply by bassamanator** [reply]
If `pause_on_runout: True` and your filament sensor is working as expected, what you're saying isn't possible unless you've made changes to the config.

I just tested the procedure I previously mentioned and it worked as expected.

BTW is your printer able to `beep`?

I would try pulling the config from the repo and starting from scratch.

**Comment by toeachbyeach** [comment]
Thankyou for all your help. I'll do a bit of a reset. Possibly altered something early on without realising and now its causing problems. 

Thanks again!
